subsys/net/lib/coap_sock/coap_sock.c

```diff
@@ -0,0 +1,1402 @@
+			   struct coap_block_context *ctx)
+{
+	int r, block1, block2, size1, size2;
+
+	block1 = get_block_option(cpkt, COAP_OPTION_BLOCK1);
+	block2 = get_block_option(cpkt, COAP_OPTION_BLOCK2);
+	size1 = get_block_option(cpkt, COAP_OPTION_SIZE1);
+	size2 = get_block_option(cpkt, COAP_OPTION_SIZE2);
+
+	size1 = size1 == -ENOENT ? 0 : size1;
+	size2 = size2 == -ENOENT ? 0 : size2;
+
+	if (is_request(cpkt)) {
+		r = update_control_block2(ctx, block2, size2);
+		if (r) {
+			return r;
+		}
+
+		return update_descriptive_block(ctx, block1, size1);
+	}
+
+	r = update_control_block1(ctx, block1, size1);
+	if (r) {
+		return r;
+	}
+
+	return update_descriptive_block(ctx, block2, size2);
+}
+
+size_t coap_next_block(const struct coap_packet *cpkt,
+		       struct coap_block_context *ctx)
+{
+	int block;
+
+	if (is_request(cpkt)) {
+		block = get_block_option(cpkt, COAP_OPTION_BLOCK1);
+	} else {
+		block = get_block_option(cpkt, COAP_OPTION_BLOCK2);
+	}
+
+	if (!GET_MORE(block)) {
+		return 0;
+	}
+
+	ctx->current += coap_block_size_to_bytes(ctx->block_size);
+
+	return ctx->current;
+}
+
+int coap_pending_init(struct coap_pending *pending,
+		      const struct coap_packet *request,
+		      const struct sockaddr *addr)
+{
+	memset(pending, 0, sizeof(*pending));
+
+	pending->id = coap_header_get_id(request);
+
+	memcpy(&pending->addr, addr, sizeof(*addr));
+
+	pending->data = request->data;
+	pending->len = request->offset;
+
+	return 0;
+}
+
+struct coap_pending *coap_pending_next_unused(
+	struct coap_pending *pendings, size_t len)
+{
+	struct coap_pending *p;
+	size_t i;
+
+	for (i = 0, p = pendings; i < len; i++, p++) {
+		if (p->timeout == 0) {
+			return p;
+		}
+	}
+
+	return NULL;
+}
+
+struct coap_reply *coap_reply_next_unused(
+	struct coap_reply *replies, size_t len)
+{
+	struct coap_reply *r;
+	size_t i;
+
+	for (i = 0, r = replies; i < len; i++, r++) {
+		if (!r->reply) {
+			return r;
+		}
+	}
+
+	return NULL;
+}
+
+static inline bool is_addr_unspecified(const struct sockaddr *addr)
+{
+	if (addr->sa_family == AF_UNSPEC) {
+		return true;
+	}
+
+	if (addr->sa_family == AF_INET6) {
+		return net_is_ipv6_addr_unspecified(
+			&(net_sin6(addr)->sin6_addr));
+	} else if (addr->sa_family == AF_INET) {
+		return net_sin(addr)->sin_addr.s4_addr32[0] == 0;
+	}
+
+	return false;
+}
+
+struct coap_observer *coap_observer_next_unused(
+	struct coap_observer *observers, size_t len)
+{
+	struct coap_observer *o;
+	size_t i;
+
+	for (i = 0, o = observers; i < len; i++, o++) {
+		if (is_addr_unspecified(&o->addr)) {
+			return o;
+		}
+	}
+
+	return NULL;
+}
+
+struct coap_pending *coap_pending_received(
+	const struct coap_packet *response,
+	struct coap_pending *pendings, size_t len)
+{
+	struct coap_pending *p;
+	u16_t resp_id = coap_header_get_id(response);
+	size_t i;
+
+	for (i = 0, p = pendings; i < len; i++, p++) {
+		if (!p->timeout) {
+			continue;
+		}
+
+		if (resp_id != p->id) {
+			continue;
+		}
+
+		return p;
```

Comment:
Is there a reason why the "coap_pending_clear(p);" was removed here?

Reply:
Yes, CoAP lib doesn't how that buffer is allocated. So it's caller responsibility to free that.
I changed in samples(coap_server) like:

k_free(pending->data);
coap_pending_clear(pending);